subsys/audio_module/audio_module.h

```diff
@@ -0,0 +1,463 @@
+ */
+struct audio_module_context;
+
+/**
+ * @brief Module's private configuration.
+ */
+struct audio_module_configuration;
+
+/**
+ * @brief Callback function for a response to a data_send as
+ *        supplied by the module user.
+ *
+ * @param handle[in/out]  The handle of the module that sent the data message
+ * @param block[in]       The audio block to operate on
+ */
+typedef void (*audio_module_response_cb)(struct audio_module_handle_private *handle,
+					 struct audio_data *block);
+
+/**
+ * @brief Private pointer to a module's functions.
+ */
+struct audio_module_functions {
+	/**
+	 * @brief Function for opening a module.
+	 *
+	 * @param handle[in/out]     The handle to the module instance
+	 * @param configuration[in]  Pointer to the module's current configuration
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*open)(struct audio_module_handle_private *handle,
+		    struct audio_module_configuration *configuration);
+
+	/**
+	 * @brief Function to close an open module.
+	 *
+	 * @param handle[in/out]  The handle to the module instance
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*close)(struct audio_module_handle_private *handle);
+
+	/**
+	 * @brief Function to configure a module.
+	 *
+	 * @param handle[in/out]     The handle to the module instance
+	 * @param configuration[in]  Pointer to the module's configuration to set
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*configuration_set)(struct audio_module_handle_private *handle,
+				 struct audio_module_configuration *configuration);
+
+	/**
+	 * @brief Function to get the configuration of a module.
+	 *
+	 * @param handle[in]          The handle to the module instance
+	 * @param configuration[out]  Pointer to the module's current configuration
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*configuration_get)(struct audio_module_handle_private *handle,
+				 struct audio_module_configuration *configuration);
+
+	/**
+	 * @brief Start a module.
+	 *
+	 * @param handle[in/out]  The handle for the module to start
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*start)(struct audio_module_handle_private *handle);
+
+	/**
+	 * @brief Stop a module.
+	 *
+	 * @param handle[in/out]  The handle for the module to stop
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*stop)(struct audio_module_handle_private *handle);
+
+	/**
+	 * @brief The core data processing function for the module. Can be either an
+	 *		  input, output or processor module type.
+	 *
+	 * @param handle[in/out]  The handle to the module instance
+	 * @param block_rx[in]    Pointer to the input audio block or NULL for an input module
+	 * @param block_tx[out]   Pointer to the output audio block or NULL for an output module
+	 *
+	 * @return 0 if successful, error otherwise
+	 */
+	int (*data_process)(struct audio_module_handle_private *handle, struct audio_data *block_rx,
```

Comment:
For my own learning, what happens if the module is async? Is then `audio_module_response_cb`  called?